Refactor Named Matches to use the lucene Matches API

romseygeek opened the pull request:
This commit reimplements the named matches API, wrapping queries that
have been given names in a new NamedQuery or NamedSpanQuery filter
class, and then extracting references to these names from the resulting
Matches objects.

romseygeek committed:
Upgrade lucene; cut named queries over to NamedMatches

romseygeek committed:
Hanging somewhere still

romseygeek committed:
Upgrade lucene (#45721)

Upgrade to a lucene snapshot that includes Matches.getSubMatches()

romseygeek committed:
Build Weight once; pass through cache

romseygeek committed:
minor revert

romseygeek committed:
precommit

romseygeek committed:
licenses

romseygeek committed:
Extract nested queries and use when in nested context

romseygeek committed:
Use hitsExecute

romseygeek committed:
Merge remote-tracking branch 'origin/feature-matches' into feature-matches

romseygeek committed:
Merge remote-tracking branch 'origin/feature-matches' into named-matches

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
feedback; test fix

romseygeek committed:
Use QueryVisitor to extract nested queries

elasticmachine commented:
Pinging @elastic/es-search (:Search/Highlighting)

romseygeek commented on server/src/test/java/org/elasticsearch/search/aggregations/metrics/TopHitsIT.java:
```diff
@@ -904,17 +904,14 @@ public void testNestedFetchFeatures() {
-        // If top_hits would have a query option then we can explain that query
         Explanation explanation = searchHit.getExplanation();
-        assertFalse(explanation.isMatch());
+        assertTrue(explanation.isMatch());
```
A nice side effect of this change is that, because we're finding relevant child queries when we build the SubSearchContext, explanations will now work.

romseygeek committed:
Add null check

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/TopHitsAggregatorFactory.java:
```diff
@@ -118,8 +124,41 @@ public Aggregator createInternal(SearchContext searchContext,
         if (highlightBuilder != null) {
             subSearchContext.highlight(highlightBuilder.build(searchContext.getQueryShardContext()));
         }
+        if (isNested(parent) && searchContext.parsedQuery() != null) {
```
I am not sure we should do that, we have the same problem on clauses that appear under a `nested` query. If the fetch phase is run inside `inner_hits` we'll break the matching. I am in favor of documenting this behavior and maybe add an option to consider each named query as independent like we do today. I think it's an improvement to consider that a named query matches only if the other required clauses are matching but we cannot handle multi-level easily.

jimczi reviewed (changes requested):
Thanks @romseygeek . I left a comment regarding `nested` fields, I think we should have a boolean to control whether we apply the query from the top and iterate over the matching named queries or if we separate each named queries and run them as should clauses independently of the other clauses. This would complicate a bit the API but I don't see a consistent way of handling `nested` and multi-level matching seamlessly.

romseygeek committed:
cleanup

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
Add matches framework

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
WIP: still some test failures

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
integ tests

romseygeek committed:
Apparently we can have null parsedquery, who knew?

romseygeek committed:
precommit

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
Remove ParsedQuery

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
Add old matching phase back, to preserve BWC behaviour for now

romseygeek committed:
imports

romseygeek committed:
test fix

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
cleanup

romseygeek committed:
null check in SubSearchContext

romseygeek committed:
Don't create match_details for completion tests

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
searchresponsetests

romseygeek committed:
remove unused method

romseygeek committed:
Merge remote-tracking branch 'origin/master' into named-matches

romseygeek committed:
tests

romseygeek commented:
I've reworked things here:  `MatchedQueriesPhase` stays in place, and I've added a new `MatchesFetchSubPhase` that generates a `Matches` object for the search query and passes it to a configurable set of `MatchesProcessors`.  I've kept the reimplementation of query names using a `NamedQuery` wrapper, and search contexts now use query visitors to pull all of these, which allows us to remove `ParsedQuery`.  We have two processors implemented, `NamedQueries` which reproduces the behaviour of `MatchedQueriesPhase` with improved accuracy, and `MatchingTerms` which reports which terms matched in a document.  `MatchingTerms` can also report matching terms per named query; in a future PR we can add a flag to disable the `MatchedQueriesPhase` when named queries are present if users just want to use the new processors.  I removed the hacky nested and inner hits query extraction code - in future we can add the ability to define a separate query to use in the `MatchesFetchSubPhase` which would let users define how they want to highlight nested or inner hits.

No docs or yaml tests yet, I can start working on those if @jimczi likes the way this is implemented.

romseygeek committed:
Handle MatchNoDocs rewrites

romseygeek committed:
test

jimczi reviewed (commented):
I like the new MatchingTerms a lot and I think this is going in the right direction. 
I have some questions/thoughts that I'd like to resolve, some yml tests or simple examples of the new API in the description would also be helpful to understand correctly.

> We have two processors implemented, NamedQueries which reproduces the behaviour of MatchedQueriesPhase with improved accuracy, and MatchingTerms which reports which terms matched in a document. MatchingTerms can also report matching terms per named query; in a future PR we can add a flag to disable the MatchedQueriesPhase when named queries are present if users just want to use the new processor

I wonder why we need to keep the two. The only difference is in the information that is reported so we could start changing the output in 8 and retain the bwc in 7.x by returning the current format by default  ? So to be clear, why don't we have a single MatchesSubPhase that adds the support for returning terms that match in a backward compatible way ?

> in future we can add the ability to define a separate query to use in the MatchesFetchSubPhase which would let users define how they want to highlight nested or inner hits.

I don't think `inner_hits` should differ from its current meaning. Currently we retrieve all `inner_hits` if the root document matches so we are expecting to highlight no matter what the state of the parent clause is. I wonder if the confusion comes from the fact that we're using the original query to highlight `nested` inner hits. This is wrong, we should be able to pass the wrapped query (the nested one) in the context so that we don't expect consumer to be aware of the level we are. For nested highlighting we should only deal with queries that targets the nested level.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

romseygeek commented:
Still something we want to do, but this PR is sufficiently out-of-date that we may as well close it and start again.
